New Sphinx tutorial, part II (#9355)
sphinx-doc/sphinx: doc/tutorial/index.rst

for development, since you will use *Python virtual environments* to create the
project.

> .. contents:: Contents

Reviewer: Is this needed? We already have local ToC table on right side bar. No reason to show it, I think.

Author: Fair enough. I added it because it looked cool and I didn't know it was possible before I started writing :)

   (.venv) $ sphinx-build -b html docs/source/ docs/build/html

And finally, open ``docs/build/html/index.html`` in your browser.  You should see
something like this:

.. figure:: /_static/tutorial/lumache-first-light.png
   :width: 80%
   :align: center
   :alt: Freshly created documentation of Lumache

   Freshly created documentation of Lumache

There we go! You created your first HTML documentation using Sphinx.

First steps to document your project using Sphinx
-------------------------------------------------

> Converting your documentation to HTML

Reviewer 2: See next comment about "export" vs. "build" vs. "compile". "Converting" should probably change accordingly.

Author: Addressed, thanks!

.. code-block:: console

   (.venv) $ make epub

After that, you will see the files corresponding to the e-book under
``docs/build/epub/``.  You can either open ``Lumache.epub`` with an
EPUB-compatible e-book viewer, like `Calibre <https://calibre-ebook.com/>`_,
or preview ``index.xhtml`` on a web browser.

.. note::

   To quickly display a complete list of possible output formats, plus some
   extra useful commands, you can run :code:`make help`.

Each output format has some specific configuration options that you can tune,
:ref:`including EPUB <epub-options>`.  For instance, the default value of
:confval:`epub_show_urls` is ``inline``, which means that, by default, URLs are
shown right after the corresponding link, in parentheses.  You can change that
behavior by adding the following code at the end of your ``conf.py``:

.. code-block:: python

   # EPUB options
   epub_show_urls = 'footnote'

With this configuration value, and after running ``make epub`` again, you will
notice that URLs appear now as footnotes, which avoids cluttering the text.
Sweet!

.. note::

   Generating a PDF using Sphinx can be done running ``make latexpdf``,

>    provided that the system has a working :math:`\LaTeX` installation,

Reviewer 3: This breaks PDF production, as `\(\LaTeX\)` mark-up is invalid. `\LaTeX` can not be used in math mode.

Author: Woops, sorry. Let's continue the conversation in #9475 and #9476

that the ``toctree`` gets rendered as a list of hyperlinks, and this allows you
to navigate to the new page you just created.  Neat!

.. warning::

   Documents outside a *toctree* will result in ``WARNING: document isn't
   included in any toctree`` messages during the build process, and will be
   unreachable for users.

Adding cross-references
~~~~~~~~~~~~~~~~~~~~~~~

One powerful feature of Sphinx is the ability to seamlessly add
:ref:`cross-references <xref-syntax>` to specific parts of the documentation:
a document, a section, a figure, a code object, etc.  This tutorial is full of
them!

To add a cross-reference, write this sentence right after the
introduction paragraph in ``index.rst``:

.. code-block:: rst
   :caption: docs/source/index.rst

   Check out the :doc:`usage` section for further information.

The :rst:role:`doc` role you used automatically references a specific document
in the project, in this case the ``usage.rst`` you created earlier.

Alternatively, you can also add a cross-reference to an arbitrary part of the
project. For that, you need to use the :rst:role:`ref` role, and add an
explicit *label* that acts as `a target`__.

> __ https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#hyperlink-targets

Author: This causes `tox -e docslint` to fail. I don't know of any way to prevent that.

Reviewer: Oops. Additional rule is needed for this case: https://github.com/sphinx-doc/sphinx/pull/9405

Author: Merged `4.x` on this branch to include #9405, thanks!

After running this command, you will see that ``index.html`` reflects the new
changes!

Exporting your documentation to other formats
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Sphinx supports a variety of formats apart from HTML, including PDF, EPUB,
:ref:`and more <builders>`.  For example, to build your documentation as an
e-book in EPUB format, run this command from the ``docs`` directory:

.. code-block:: console

>    (.venv) $ make epub

Reviewer 4: Is there a reason we're doing epub here instead of `pseudoxml`? Just ease of browsing?

Author: Yes, I went back to your original suggestion of using EPUB, because it's a more familiar format. For beginners, it's unclear what the `pseudoxml` builder is for, and in fact:

> This builder is used for debugging the Sphinx/Docutils “Reader to Transform to Writer” pipeline.

https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.xml.PseudoXMLBuilder

The good thing is that, even if the user doesn't have an EPUB viewer, they can still open the index in a web browser.

:confval:`epub_show_urls` is ``inline``, which means that, by default, URLs are
shown right after the corresponding link, in parentheses.  You can change that
behavior by adding the following code at the end of your ``conf.py``:

.. code-block:: python

   # EPUB options
   epub_show_urls = 'footnote'

With this configuration value, and after running ``make epub`` again, you will
notice that URLs appear now as footnotes, which avoids cluttering the text.
Sweet!

.. warning::

   Generating a PDF using Sphinx can be done running ``make latexpdf``,
   provided that the system has a working :math:`\LaTeX` installation,
   as explained in the documentation of :class:`sphinx.builders.latex.LaTeXBuilder`.
   Although this is perfectly feasible, such installations are often big,
   and in general LaTeX requires careful configuration in some cases,
   so PDF generation is out of scope for this tutorial.

More Sphinx customization using extensions
------------------------------------------

Enabling a built-in extension
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In addition to these configuration values, you can customize Sphinx even more

> by using :doc:`extensions <../usage/extensions/index>`.  Sphinx ships several

Reviewer 4: Should we use absolute links for these, so that if this content moves we don't have to update them. (Thinking we might break the tutorial up into multiple pages in the future, though I guess they'll still be in the `tutorial/` dir)

Author: Good idea, changing it

:ref:`built-in ones <built-in-extensions>`, and there are many more
:ref:`maintained by the community <third-party-extensions>`.

For example, to enable the :mod:`sphinx.ext.duration` extension,
locate the ``extensions`` list in your ``conf.py`` and add one element as
follows:

.. code-block:: python

   # Add any Sphinx extension module names here, as strings. They can be
   # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
   # ones.
   extensions = [
       'sphinx.ext.duration',
   ]

After that, every time you generate your documentation, you will see a short
durations report at the end of the console output, like this one:

.. code-block:: console

   (.venv) $ make html
   ...
   The HTML pages are in build/html.

   ====================== slowest reading durations =======================

>    0.042 temp/source/index

Reviewer 4: Great example 👍

Using a third-party theme
~~~~~~~~~~~~~~~~~~~~~~~~~

Themes, on the other hand, are a particular class of extensions that allow you

> to customize the appearance of your documentation.  Again, Sphinx has several

Reviewer 4: ```suggestion
to customize the appearance of your documentation.  Sphinx has several
```

> :ref:`built-in themes <builtin-themes>`, and there are also third-party ones.

Reviewer 4: Link to https://sphinx-themes.org/ for third-party?

Author: Good idea 💪🏽

For example, to use the `Furo <https://pradyunsg.me/furo/>`_ third-party theme

> in your HTML documentation, first you will need to install it with ``pip`` on

Reviewer 4: ```suggestion
in your HTML documentation, first you will need to install it with ``pip`` in
```

Reviewer 4: Seems we're also noting that themes only apply to HTML docs here implicitly, I wonder if this should be more explicit?

Author: I wonder if we should address that in a separate PR, since all the documentation seems to be suggesting that. Every "theme" reference in https://www.sphinx-doc.org/en/master/contents.html is for "HTML" ("HTML Theming", "HTML theme development", "Documentation using a custom theme or integrated in a website"). While writing this response, I tried searching for "sphinx pdf theme", and I got [a Stack Overflow answer saying "this can't be done"](https://stackoverflow.com/a/9806839/554319), an [unhelpful thread in sphinx-users](https://groups.google.com/g/sphinx-users/c/LBGxmGESUhg), and finally a [quite new PDF theme](https://pypi.org/project/sphinx-business-theme/) that however doesn't even use LaTeX, but [weasyprint](https://pypi.org/project/weasyprint/) (¡hola @nekmo! 👋🏽)

Even if it's possible to have Sphinx themes for things other than HTML, I would like to not point people to things they can't easily find at the moment, without rethinking other parts of the Sphinx documentation first.

Author: In any case, I changed the header to say "HTML theme" to make room for other kind of themes in the future 👍🏽

your Python virtual environment, like this:

.. code-block:: console

   (.venv) $ pip install furo

And then, locate the ``html_theme`` variable on your ``conf.py`` and replace
its value as follows:

.. code-block:: python

   # The theme to use for HTML and HTML Help pages.  See the documentation for
   # a list of builtin themes.
   #

>    html_theme = 'alabaster'

Reviewer 5: Should this be `furo`?

Author: Woops good catch!

With this change, you will notice that your HTML documentation has now a new
appearance:

.. figure:: /_static/tutorial/lumache-furo.png
   :width: 80%
   :align: center
   :alt: HTML documentation of Lumache with the Furo theme

   HTML documentation of Lumache with the Furo theme

Narrative documentation in Sphinx
---------------------------------

> Inserting documents in the project hierarchy

Reviewer 4: I wonder about a more approachable term here. I feel like `hierarchy` isn't really a known concept in terms of documents. Perhaps something like "Structuring your documentation across multiple pages", "Building a menu in the sidebar" or something more goal-oriented.

Author: +100 for "something more goal-oriented" 👍🏽 your first suggestion looks good to me. Renamed the header.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> As discussed at the beginning, ``index.rst`` is the :term:`master document`,

Reviewer 4: I don't love references to earlier content like this (maybe I don't remember it being discussed? Maybe someone linked me directly to this section for some reason?) -- but don't feel too strongly.

Reviewer 4: (I also wonder about renaming `master` to `primary` or `index`, but that's another discussion :)

Author: You're right. (As I read the other day: "Never assume that anyone would read your article word to word, from beginning to end".) Will rephrase it.

About "master document", yeah, that's for another day :)

whose main function is to serve as a welcome page and to contain the root of
the "table of contents tree" (or *toctree*).  Sphinx allows you to assemble
a project from different files, which is helpful when the project grows.

As an example, create a new file ``docs/source/usage.rst`` (next to
``index.rst``) with these contents:

.. code-block:: rest

>    Usage

Reviewer 4: ```suggestion
   # file: docs/source/usage.rst
   
   Usage
```

I find this useful to be doubly explicit. I've seen some themes have a way to actually pull this out into a "header" on the codeblock, which could be a nice benefit in the future. 

Would want to be consistent with this across all the file examples, but might be a good idea.

Reviewer 4: Actually, a `caption` might be the right solution here: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-option-code-block-caption

Author: Nice! Adding this to all files

   =====

   Installation
   ------------

   To use Lumache, first install it using pip:

   .. code-block:: console

      (.venv) $ pip install lumache

This new file contains two :ref:`section <rst-sections>` headers, normal
paragraph text, and a ``code-block`` directive that renders a block of content
as source code, with appropriate syntax highlighting (in this case, generic
``console`` text).

.. note::

   You can read `the list of available highlight

>    languages <https://pygments.org/docs/lexers/>`_ in the Pygments

Reviewer 4: This is super overwhelming.. I wonder if it's not worth including this in the tutorial, but adding it to the reference docs somewhere?

Author: True, I added this to the `code-block` documentation instead 👍🏽

   documentation.

The structure of the document is determined by the succession of heading
styles, which means that, by using ``---`` for the "Installation" section
after ``===`` for the "Usage" section, you have declared "Installation" to
be a *subsection* of "Usage".

> In addition, add a ``toctree`` :ref:`directive <rst-directives>` at the end

Reviewer 4: ```suggestion
To complete the process, you can add a ``toctree`` :ref:`directive <rst-directives>` at the end
```

A bit more clear that this is the "second step" of the toctree process.

of ``index.rst`` including the document you just created,
as follows:

.. code-block:: rest

   Contents
   --------

   .. toctree::

      usage

This step inserts that document in the root of the *toctree*, so now it belongs
to the structure of your project.  If you export the documentation to HTML
running ``make html``, you will see that the ``toctree`` gets rendered as a
list of hyperlinks, and this allows you to navigate to the new page you

> just created.  Neat!

Reviewer 4: I wonder if having a tree diagram for the TOC is useful here -- to help visually show the structure that was created. I find people always get confused with this:

```
index
 - usage
```

Author: Added a small text representation of the structure.

.. warning::

   Every document should belong to a *toctree*.  Otherwise, Sphinx will emit a
   ``WARNING: document isn't included in any toctree``, and the end result
   will depend on the builder.  For the HTML builder, the page will not be
   linked from anywhere (therefore it will not be discoverable), whereas for

>    the PDF builder, it will not be included at all.

Reviewer 4: This warning feels a bit esoteric and complex. I'd remove the PDF section at least, or a generic warning "documents outside the toctree will be impossible for many users to find" or something. 

I'd love to see an "error messages" guide in the docs that have this kind of very explicit content though. Would be great for new users who google errors.

Author: True, it's too long. Will rephrase it to make it simpler.

Author: About the "errors" section, I think the main issue is that `make ...` produces lots of colored output, and warnings that should be addressed pass silently. It's a difficult problem I think, but the case of out-of-doctree documents should be treated as an error in my opinion (with an option to ignore it).

Author: (I'm 👍🏽 on adding a FAQ some day)

Adding cross-references
~~~~~~~~~~~~~~~~~~~~~~~

One powerful feature of Sphinx is the ability to seamlessly add
:ref:`cross-references <xref-syntax>` to specific parts of the documentation:
a document, a section, a figure, a code object, etc.  This tutorial is full of
them!

> To add a straightforward cross-reference, write this sentence right after the

Reviewer 4: ```suggestion
To add a cross-reference, write this sentence right after the
```

introduction paragraph in ``index.rst``:

.. code-block:: rest

   Check out the :doc:`usage` section for further information.

The :rst:role:`doc` role you used automatically references a specific document
in the project, in this case the ``usage.rst`` you created earlier.

Alternatively, you can also add a cross-reference to an arbitrary part of the
project. For that, you need to use the :rst:role:`ref` role, and add an
explicit *label* that can act as a target.

For example, to reference the "Installation" subsection, add a label right
before the heading, as follows:

.. code-block:: rest

> 

Reviewer 4: ```suggestion
     :emphasize-lines: 4

```

This might be useful.

Author: Love this! Will use it more extensively

After running this command, you will see that ``index.html`` reflects the new
changes!

Exporting your documentation to other formats
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Sphinx supports a variety of formats apart from HTML, including PDF, EPUB,
:ref:`and more <builders>`.  For example, to build your documentation as an
e-book in EPUB format, run this command from the ``docs`` directory:

.. code-block:: console

   (.venv) $ make epub

After that, you will see the files corresponding to the e-book under

> ``docs/build/html/``.  You can either open ``Lumache.epub`` with an

Reviewer 6: ```suggestion
``docs/build/epub/``.  You can either open ``Lumache.epub`` with an
```

I think the directory is build/epub.

Author: Oops, good catch! Thanks @humitos 🙏🏽

EPUB-compatible e-book viewer, like `Calibre <https://calibre-ebook.com/>`_,
or preview ``index.xhtml`` on a web browser.

.. note::

   To quickly display a complete list of possible output formats, plus some
   extra useful commands, you can run :code:`make help`.

Each output format has some specific configuration options that you can tune,
:ref:`including EPUB <epub-options>`.  For instance, the default value of
:confval:`epub_show_urls` is ``inline``, which means that, by default, URLs are
shown right after the corresponding link, in parentheses.  You can change that
behavior by adding the following code at the end of your ``conf.py``:

.. code-block:: python

   # EPUB options
   epub_show_urls = 'footnote'

With this configuration value, and after running ``make epub`` again, you will
notice that URLs appear now as footnotes, which avoids cluttering the text.
Sweet!

> .. warning::

Reviewer 2: ```suggestion
.. note::
```
A warning seems a bit too strong to me.

Author: You're right, accepting

   Generating a PDF using Sphinx can be done running ``make latexpdf``,
   provided that the system has a working :math:`\LaTeX` installation,
   as explained in the documentation of :class:`sphinx.builders.latex.LaTeXBuilder`.
   Although this is perfectly feasible, such installations are often big,
   and in general LaTeX requires careful configuration in some cases,
   so PDF generation is out of scope for this tutorial.

> More Sphinx customization using extensions

Reviewer 2: ```suggestion
More Sphinx customization
```
See comment below on HTML themes.

Author: Done locally

------------------------------------------

> 

Reviewer 2: I suggest writing a few sentences here to introduce the different kinds of customization.

Author: Done! (Added a brief introductory sentence)

Enabling a built-in extension
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In addition to these configuration values, you can customize Sphinx even more
by using :doc:`extensions </usage/extensions/index>`.  Sphinx ships several
:ref:`built-in ones <built-in-extensions>`, and there are many more
:ref:`maintained by the community <third-party-extensions>`.

For example, to enable the :mod:`sphinx.ext.duration` extension,
locate the ``extensions`` list in your ``conf.py`` and add one element as
follows:

.. code-block:: python
   :caption: docs/source/conf.py

   # Add any Sphinx extension module names here, as strings. They can be
   # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
   # ones.
   extensions = [
       'sphinx.ext.duration',
   ]

After that, every time you generate your documentation, you will see a short
durations report at the end of the console output, like this one:

.. code-block:: console

   (.venv) $ make html
   ...
   The HTML pages are in build/html.

   ====================== slowest reading durations =======================
   0.042 temp/source/index

Using a third-party HTML theme
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Themes, on the other hand, are a particular class of extensions that allow you

Reviewer 2: While a theme is an extension in the abstract sense, I think the term "extension" for Sphinx is mostly used in a technical sense about a Python module which must be added to the ``extensions`` conf var.
(ping @tk0miya for potential corrections of understanding)

Author: According to the Glossary, "A custom role, directive or other aspect of Sphinx that allows users to modify any aspect of the build process within Sphinx." https://www.sphinx-doc.org/en/master/glossary.html#term-extension

And then, according to https://www.sphinx-doc.org/en/master/usage/extensions/index.html, "Sphinx allows adding “extensions” to the build process, each of which can modify almost any aspect of document processing.".

In any case, I get your point: that themes are not extensions, but another form of customization. Will rephrase this.

ones <https://sphinx-themes.org/>`_.

For example, to use the `Furo <https://pradyunsg.me/furo/>`_ third-party theme
in your HTML documentation, first you will need to install it with ``pip`` in
your Python virtual environment, like this:

.. code-block:: console

   (.venv) $ pip install furo

And then, locate the ``html_theme`` variable on your ``conf.py`` and replace
its value as follows:

.. code-block:: python
   :caption: docs/source/conf.py

   # The theme to use for HTML and HTML Help pages.  See the documentation for
   # a list of builtin themes.
   #
   html_theme = 'furo'

With this change, you will notice that your HTML documentation has now a new
appearance:

.. figure:: /_static/tutorial/lumache-furo.png
   :width: 80%
   :align: center
   :alt: HTML documentation of Lumache with the Furo theme

   HTML documentation of Lumache with the Furo theme

Narrative documentation in Sphinx
---------------------------------

Structuring your documentation across multiple pages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The file ``index.rst`` created by ``sphinx-quickstart`` is the :term:`master

> document`, whose main function is to serve as a welcome page and to contain the

Reviewer 2: In 4.0 "master" was changed to "root", https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-root_doc

Author: Ah, great to know! Changing in the rest of the document as well

The :rst:role:`doc` role you used automatically references a specific document
in the project, in this case the ``usage.rst`` you created earlier.

Alternatively, you can also add a cross-reference to an arbitrary part of the
project. For that, you need to use the :rst:role:`ref` role, and add an
explicit *label* that can act as a target.

For example, to reference the "Installation" subsection, add a label right
before the heading, as follows:

.. code-block:: rest
   :caption: docs/source/usage.rst
   :emphasize-lines: 4

   Usage
   =====

   .. _installation:

   Installation
   ------------

   ...

And make the sentence you added in ``index.rst`` look like this:

.. code-block:: rest
   :caption: docs/source/index.rst

   Check out the :doc:`usage` section for further information, including how to
   :ref:`install <installation>` the project.

Notice a trick here: the ``install`` part specifies how the link will look like
(we want it to be a specific word, so the sentence makes sense), whereas the
``<installation>`` part refers to the actual label we want to add a
cross-reference to. Both the ``:doc:`` and the ``:ref:`` roles will be rendered

> as hyperlinks in the HTML documentation.

Reviewer 2: Is this the first time the ``:ref:`` role is explained? If so I think it would be good to explain that the title is optional, and what happens when none is given.

Author: Good idea! Changed

After running this command, you will see that ``index.html`` reflects the new
changes!

> Exporting your documentation to other formats

Reviewer: Is it better to use "export" instead of "build"? AFAIK, Sphinx is usually described as a "documentation builder". So it would be better to use "build". Personally, I still don't understand what is better wording to this behavior; build, convert or others (Because I'm not a native speaker!).

Reviewer 2: Good point. I think it is worthwhile being consistent with the existing interface and use the term "building" (or whichever variantion fits the context): a new format is added by creating a [Builder](https://www.sphinx-doc.org/en/master/usage/builders/index.html) extension and you select the output format with ``-b`` on the command line.
As another alternative to "export" I suggest "compile". To me "export" seems more like a direct write in a different format, but Sphinx really is performing tasks akin to an ordinary compiler.

Author: Good points, I will rephrase to "build". "Compile" is too attached to machine instructions, I think.

Author: I was tempted to introduce a preposition that's probably incorrect ("build your docs AS html", "build your docs TO html"), so instead I kept "build your HTML docs".

Author: (Forget my last comment - sometimes I think it makes sense to say "build docs AS x")

   :caption: docs/source/usage.rst

   Usage
   =====

   Installation
   ------------

   To use Lumache, first install it using pip:

   .. code-block:: console

      (.venv) $ pip install lumache

This new file contains two :ref:`section <rst-sections>` headers, normal
paragraph text, and a :ref:`code-block <rst-code-block>` directive that renders
a block of content as source code, with appropriate syntax highlighting
(in this case, generic ``console`` text).

The structure of the document is determined by the succession of heading
styles, which means that, by using ``---`` for the "Installation" section
after ``===`` for the "Usage" section, you have declared "Installation" to
be a *subsection* of "Usage".

To complete the process, add a ``toctree`` :ref:`directive <rst-directives>` at
the end of ``index.rst`` including the document you just created, as follows:

.. code-block:: rest
   :caption: docs/source/index.rst

   Contents
   --------

   .. toctree::

      usage

This step inserts that document in the root of the *toctree*, so now it belongs
to the structure of your project, which so far looks like this:

.. code-block:: text

   index
   └── usage

> If you export the documentation to HTML running ``make html``, you will see

Reviewer: export?

Author: ✔️

that the ``toctree`` gets rendered as a list of hyperlinks, and this allows you
to navigate to the new page you just created.  Neat!

.. warning::

   Documents outside a *toctree* will result in ``WARNING: document isn't
   included in any toctree`` messages during the build process, and will be
   unreachable for users.

Adding cross-references
~~~~~~~~~~~~~~~~~~~~~~~

One powerful feature of Sphinx is the ability to seamlessly add
:ref:`cross-references <xref-syntax>` to specific parts of the documentation:
a document, a section, a figure, a code object, etc.  This tutorial is full of
them!

To add a cross-reference, write this sentence right after the
introduction paragraph in ``index.rst``:

.. code-block:: rest
   :caption: docs/source/index.rst

   Check out the :doc:`usage` section for further information.

The :rst:role:`doc` role you used automatically references a specific document
in the project, in this case the ``usage.rst`` you created earlier.

Alternatively, you can also add a cross-reference to an arbitrary part of the
project. For that, you need to use the :rst:role:`ref` role, and add an

> explicit *label* that can act as a target.

Reviewer: FYI: We've called this notation "label" longly. But it has been called "Hyperlink targets" notation formally. So it's natural to act as a target :-)
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#hyperlink-targets

Author: Thanks for the clarification :) Will use this opportunity to link to the reST docs.

For example, to reference the "Installation" subsection, add a label right
before the heading, as follows:

> .. code-block:: rest

Reviewer: We usually uses `rst` as a highlight language for reStructuredText.

```
tkomiya@Alrescha> grep -r code-block:: doc | grep ':: rst'
doc/man/sphinx-autogen.rst:.. code-block:: rst
doc/development/theming.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/helloworld.rst:.. code-block:: rst
doc/development/tutorials/helloworld.rst:.. code-block:: rst
doc/development/tutorials/recipe.rst:.. code-block:: rst
doc/development/tutorials/recipe.rst:.. code-block:: rst
doc/development/tutorials/autodoc_ext.rst:.. code-block:: rst
doc/usage/configuration.rst:   .. code-block:: rst
doc/usage/extensions/doctest.rst:.. code-block:: rst
doc/faq.rst:  .. code-block:: rst
```

I noticed some code-blocks uses "rest"...
```
tkomiya@Alrescha> grep -r code-block:: doc | grep ':: rest'
doc/tutorial/index.rst:.. code-block:: rest
doc/tutorial/index.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: restructuredtext
doc/usage/quickstart.rst:.. code-block:: restructuredtext
```

Reviewer: Note: Both `rest` and `rst` are supported.
https://pygments.org/docs/lexers/#pygments.lexers.markup.RstLexer

Author: Fair enough, I will convert everything to `rst` for consistency.
